### ISSUE TITLE
TensorRT inference time issue

### ISSUE THREAD

**Issue by neo133** [ultralytics/yolov5#9231]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi,
Thanks a lot for the amazing repo.

I was checking The TensorRT export tutorial. I found that **before TensorRT** conversion the inference time for the **yolov5s** was **12.5ms** and **after TensorRT** conversion, the inference time is **11ms**. But shouldn't it be better than only this little improvement? Please suggest if I am doing anything wrong.  I have attached a screenshot for reference.

@glenn-jocher any comments please.

![Screenshot from 2022-08-31 11-45-35](https://user-images.githubusercontent.com/27700675/187607373-f68df30d-959a-4df4-9dc0-79adad7abe72.png)

Also, I have attached the results of benchmark.py

![Screenshot from 2022-08-31 12-02-31](https://user-images.githubusercontent.com/27700675/187609824-ed7f882f-7894-442f-bc39-7fc84e4d26ea.png)

**Comment by glenn-jocher** [comment]
@neo133 👋 Hello! Thanks for asking about YOLOv5 🚀 **benchmarks**. YOLOv5 inference is officially supported in 11 formats, and all formats are benchmarked for identical accuracy and to compare speed every 24 hours by the [YOLOv5 CI](https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml?query=event%3Aschedule).

💡 ProTip: Export to ONNX or OpenVINO for up to 3x CPU speedup. See [CPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6613).
💡 ProTip: Export to TensorRT for up to 5x GPU speedup. See [GPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6963).

Format                      | `export.py --include`         | Model
:---                        | --:                           | :--
[PyTorch](https://pytorch.org/)                     | -                             | `yolov5s.pt`
[TorchScript](https://pytorch.org/docs/stable/jit.html)                 | `torchscript`                 | `yolov5s.torchscript`
[ONNX](https://onnx.ai/)                        | `onnx`                        | `yolov5s.onnx`
[OpenVINO](https://docs.openvino.ai/latest/index.html)                    | `openvino`                    | `yolov5s_openvino_model/`
[TensorRT](https://developer.nvidia.com/tensorrt)                    | `engine`                      | `yolov5s.engine`
[CoreML](https://github.com/apple/coremltools)                      | `coreml`                      | `yolov5s.mlmodel`
[TensorFlow SavedModel](https://www.tensorflow.org/guide/saved_model)       | `saved_model`                 | `yolov5s_saved_model/`
[TensorFlow GraphDef](https://www.tensorflow.org/api_docs/python/tf/Graph)         | `pb`                          | `yolov5s.pb`
[TensorFlow Lite](https://www.tensorflow.org/lite)             | `tflite`                      | `yolov5s.tflite`
[TensorFlow Edge TPU](https://coral.ai/docs/edgetpu/models-intro/)         | `edgetpu`                     | `yolov5s_edgetpu.tflite`
[TensorFlow.js](https://www.tensorflow.org/js)               | `tfjs`                        | `yolov5s_web_model/`


## Benchmarks

Benchmarks below run on a Colab Pro with the YOLOv5 tutorial notebook <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>. To reproduce:
```bash
python utils/benchmarks.py --weights yolov5s.pt --imgsz 640 --device 0
```

### Colab Pro V100 GPU

```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=0, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CUDA:0 (Tesla V100-SXM2-16GB, 16160MiB)
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 46.7/166.8 GB disk)

Benchmarks complete (458.07s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623                10.19
1             TorchScript        0.4623                 6.85
2                    ONNX        0.4623                14.63
3                OpenVINO           NaN                  NaN
4                TensorRT        0.4617                 1.89
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623                21.28
7     TensorFlow GraphDef        0.4623                21.22
8         TensorFlow Lite           NaN                  NaN
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

### Colab Pro CPU

```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=cpu, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CPU
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 41.5/166.8 GB disk)

Benchmarks complete (241.20s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623               127.61
1             TorchScript        0.4623               131.23
2                    ONNX        0.4623                69.34
3                OpenVINO        0.4623                66.52
4                TensorRT           NaN                  NaN
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623               123.79
7     TensorFlow GraphDef        0.4623               121.57
8         TensorFlow Lite        0.4623               316.61
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

Good luck 🍀 and let us know if you have any other questions!

**Comment by neo133** [comment]
Hi @glenn-jocher ,
So the results I have posted are from the tutorial colab notebook that is being provided in the repo. How are you getting better TensorRT inference time than me? Is there anything wrong?

**Comment by glenn-jocher** [comment]
@neo133 hi there. The main difference I see is that we ran our benchmarks on Colab Pro which comes with 8 CPUs and a V100, while the free Colab provides 2 CPUs and a T4 as you can see in your screenshot. It might be that the low CPU count is limiting the speed. I'll try to reproduce, it would be good to provide official Colab free T4 benchmarks also for reference.

**Comment by glenn-jocher** [comment]
Ran on Colab free.
```bash
utils/benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=, half=False, test=False, pt_only=False, hard_fail=False
Checking setup...
YOLOv5 🚀 v6.2-80-g55b0096 Python-3.7.13 torch-1.12.1+cu113 CUDA:0 (Tesla T4, 15110MiB)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 43.2/78.2 GB disk)

Benchmarks complete (260.31s)
                   Format  Size (MB)  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch       14.1        0.4716                 8.98
1             TorchScript       28.1        0.4716                 7.87
2                    ONNX       28.0        0.4716                16.79
3                OpenVINO        NaN           NaN                  NaN
4                TensorRT       31.9        0.4716                 6.19
5                  CoreML        NaN           NaN                  NaN
6   TensorFlow SavedModel       27.8        0.4716                28.14
7     TensorFlow GraphDef       27.8        0.4716                27.50
8         TensorFlow Lite        NaN           NaN                  NaN
9     TensorFlow Edge TPU        NaN           NaN                  NaN
10          TensorFlow.js        NaN           NaN                  NaN
```


Well if you are running at TRT you should make sure to export at --half, which will roughly speed up your inference 2x. I think in the official benchmarks --half was enabled by default for TRT, and now it requires manually passing --half.

**Comment by neo133** [comment]
I will try and let you know.
Thank you Glenn for all the help.

**Comment by neo133** [comment]
Hi,
Thanks a lot I got approx 2X speedup...however I got another issue which I will open another one.
I'll close this one.

**Comment by glenn-jocher** [comment]
@neo133 you're welcome! I'm glad to hear you achieved the speedup. Please feel free to open another issue for any new concerns, and we'll be here to assist. 👍 Thank you for your participation in the YOLO community!
